### PR TITLE
DrivAerML: 4L/256d+T_max=10 (ultra-short cosine on winning architecture)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

T_max=10 was optimal for TandemFoil (75.59 val, best across T_max=10/15/20/30). With the newly confirmed 4L/256d architecture as the winner for DrivAerML, does T_max=10 also help DrivAerML? On TandemFoil, T_max=10 creates extremely rapid LR cycling (~75 cosine cycles/epoch at 750 steps/epoch), producing the best minima. The baseline uses T_max=30; we test whether ultra-short cycling transfers to DrivAerML.

Note: DrivAerML has ~394 steps/epoch, so T_max=10 gives ~39 cosine cycles/epoch vs ~13 cycles/epoch for T_max=30.

## Instructions

Use the new DrivAerML baseline command but with `--cosine_t_max 10`:

```bash
cd target/icml2026 && python train.py \
  --dataset drivaerml \
  --optimizer adamw \
  --lr 5e-4 \
  --cosine_t_max 10 \
  --no-use-ema \
  --enable-fourier \
  --model-layers 4 \
  --model-hidden-dim 256 \
  --model-heads 4 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200 \
  --wandb_group drivaerml-4L256d-tmax10
```

The only change vs baseline is `--cosine_t_max 10` (baseline uses 30). All other flags are identical.

**CRITICAL flags (do not omit):**
- `--batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200` — required to avoid OOM
- `--model-heads 4` — required for 256d model (256/4 = 64 per head)
- `--epochs 999` — mandatory, the timeout controls training duration

## Baseline

Current DrivAerML best (PR #2550, violet):
- **val_primary/surface_rel_l2_pct:** 12.96% (lower is better)
- **test_primary/surface_rel_l2_pct:** 14.41%
- **W&B run:** 8s5i8y06
- **Architecture:** Fourier + 4L/256d + AdamW lr=5e-4 + T_max=30 + no-EMA, 43 epochs (still converging)
- **External target:** <3.71% (AB-UPT) — 3.5x gap remaining

**Target:** beat val < 12.96%

**Reproduce baseline:**
```bash
cd target/icml2026 && python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine_t_max 30 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 256 --model-heads 4 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200
```

### CODE REVIEW

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2594 because ADVISOR: Closing — DrivAerML baseline is 5.73% (4L/384d, PR #2602). All 4L/256d experiments are obsolete. You'll receive a new 4L/384d assignment shortly..

**Comment by tcapelle** [comment]
STUDENT luffy:

## Results

### DrivAerML: 4L/256d + T_max=10 (ultra-short cosine schedule)

| Metric | Baseline (PR #2550, T_max=30) | This Run (T_max=10) | Delta |
|--------|------|------|-------|
| **val_primary/surface_rel_l2_pct** | 12.96% | **5.28%** (epoch 245) | **-7.68pt (59% relative)** |
| **test_primary/surface_rel_l2_pct** | 14.41% | **7.20%** (epoch 277) | **-7.21pt (50% relative)** |
| Epochs completed | 43 | 277 | +234 |
| Training loss (final) | — | 0.006 | — |

**W&B run:** [q4li4j9a](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/q4li4j9a)

### Command used
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 10 \
  --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 256 \
  --model-heads 4 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --wandb-group drivaerml-4L256d-tmax10 \
  --agent luffy --wandb-name "luffy/drivaerml-4L256d-tmax10"
```

### What happened

**T_max=10 is a clear winner over T_max=30 for DrivAerML.** The ultra-short cosine cycling (39 cycles/epoch with 394 steps) produced dramatically better convergence over the 180-minute budget:

- **Best val of 5.28%** (epoch 245) vs baseline 12.96% — nearly halved the error
- **Test of 7.20%** (measured at final epoch 277, not best-val checkpoint) vs baseline 14.41%
- **277 epochs completed** vs 43 for the baseline, same 180-min budget
- **External target gap narrowing:** 5.28% val vs 3.71% AB-UPT — only 1.57pt away

**Key observations:**
1. **Convergence trajectory:** Rapid initial improvement (62.6% → 12.3% in first 48 epochs), then slower gains (12.3% → 5.28% over next 200 epochs)
2. **Cosine cycle pattern:** Best val metrics don't consistently occur at cycle bottoms (LR=0) — many new bests happen at mid-cycle or even peak LR, suggesting the rapid cycling acts more as a regularizer than a traditional schedule
3. **Instability increases over time:** Val oscillation widens from ±2pt (early) to ±3-4pt (late epochs). Occasional spikes to 9-11% even after reaching 5.x% — suggests the model is in a sharp minimum with fragile generalization at high LR
4. **Still improving at cutoff:** New best at epoch 245, and several near-bests in final 30 epochs. More compute would likely yield further gains

**Why T_max=10 beats T_max=30:** With 394 steps/epoch, T_max=10 gives 39.4 cycles per epoch vs 13.1 for T_max=30. This creates a much more stochastic optimization landscape — the LR never stays high or low for long, effectively acting as warm restarts with very short cycles. This appears to help the model escape local minima more frequently.

### Peak memory usage
~3.9 GB VRAM on RTX PRO 6000 Blackwell (batch_size=1, 50k surface points)

### Suggested follow-ups
1. **T_max=5 or T_max=3:** Even shorter cycles might push further, given T_max=10 > T_max=30 > T_max=50
2. **Checkpoint at best val:** The test metric (7.20%) was measured at epoch 277, not at best val (epoch 245 at 5.28%). Adding checkpointing could capture the best model state and likely give a better test score
3. **LR reduction in later epochs:** The growing instability suggests a lower peak LR in later epochs (e.g., cosine with decay envelope) could stabilize the plateau
4. **Larger surface points (1M):** The research state notes 1M train surface points >> 50k for DrivAerML — combining T_max=10 with 1M points could be powerful
